File: JavaScript/343-Integer Break.js
/**
 * @param {number} n
 * @return {number}
 */
var integerBreak = function(n) {
    const memo = new Map([[2,1],[3,2],[4,4],[5,6],[6,9]]);
    
    let dfs = (n) =>{
        if (memo.has(n)) return memo.get(n);
        return 3*dfs(n-3);
    }
    
    return dfs(n);
};

File: C++/343-Integer Break.cpp
class Solution {
public:
    int integerBreak(int n) {
        if (n <= 3) return n-1;
        vector<int> dp(n+1, -1);
        return dfs(dp, n);
    }
    int dfs(vector<int>& dp, int n) {
        if (dp[n] != -1) return dp[n];
        int res = n;
        for (int i = 1; i < n; i++){
            res = max(res, i*dfs(dp, n-i));
        }
        return dp[n] = res;
    }
};